File: Minimum Window Substring.java
public class Solution {
    public String minWindow(String S, String T) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int []count = new int[256];
        int []track = new int[256];
        for(char c:T.toCharArray())
        {
            count[c]++; 
        }
        
        int minLen = Integer.MAX_VALUE;
        int size = 0;
        int start = 0;
        String substr = null;
        String minSubStr = null;
        for(int i = 0; i < S.length(); i++)
        {
            char key = S.charAt(i);
            track[key]++;
            if(track[key] <= count[key]&&count[key] > 0)
            {
                size++;
            }
            
            if(size==T.length())
            {
                int end = i;
                int newStart = start;
                for(int j = start; j <= end; j++)
                {
                    newStart = j;
                    int index = S.charAt(j);
                    if(track[index]==count[index])
                    {
                        substr = S.substring(j, end+1);
                        if(substr.length()<minLen)
                        {
                            minSubStr = substr;
                            minLen = minSubStr.length();
                        }
                        break;
                    }
                    if(track[index]>count[index]&&count[index]>0)
                    {
                        track[index]--;
                    }
                }
                start = newStart;
            }
        }
        
        return minLen==Integer.MAX_VALUE? "":minSubStr;
        
    }
}

File: Minimum Window Substring.cpp
class Solution {
public:
    string minWindow(string S, string T) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        string result;
        if(S.empty()||T.empty())
            return result;
        
        vector<int> record(255, 0);
        for(int i = 0; i < T.size(); i++)
        {
            record[T[i]]++;
        }
        
        int start = 0;
        int end = 0;
        
        int size = 0;
        vector<int> check(255, 0);
        int maxLen = INT_MAX;
        int maxStart = 0;
        int maxEnd = 0;
        for(int i = 0; i < S.size(); i++)
        {
            check[S[i]]++;
            if(check[S[i]] <=  record[S[i]]&&record[S[i]] > 0)
            {
                size++;
            }
            
            if(size == T.size())
            {
                end = i;
                while(start <= end)
                {
                    char c = S[start];
                    if(check[c] > record[c]&&record[c]>0)
                        check[c]--;
                    else if(check[c] == record[c] && record[c]>0)
                    {
                        int len = end - start + 1;
                        if(len < maxLen)
                        {
                            maxLen = len;
                            maxStart = start;
                            maxEnd = end;
                        }
                        break;
                    }
                    start++;
                }
            }
        }
        
        if(maxLen == INT_MAX)
            return S.substr(0,0);
        else
            return S.substr(maxStart, maxLen);
    }
